lib/executor/variable_looping_vus.go

```diff
@@ -553,68 +401,59 @@ func (vlv VariableLoopingVUs) Run(ctx context.Context, out chan<- stats.SampleCo
 		vuHandles[i] = vuHandle
 	}
 
-	rawStepEvents := lib.StreamExecutionSteps(ctx, startTime, rawExecutionSteps, true)
-	gracefulLimitEvents := lib.StreamExecutionSteps(ctx, startTime, gracefulExecutionSteps, false)
-
-	// 0 <= currentScheduledVUs <= currentMaxAllowedVUs <= maxVUs
-	var currentScheduledVUs, currentMaxAllowedVUs uint64
+	gracefulRampDownPeriod := vlv.config.GetGracefulRampDown()
+	var rampDowns uint64
+	hardStopVU := func(vuNum uint64) {
+		spent := time.Since(startTime)
+		delay := gracefulRampDownPeriod
+		if spent+delay > maxDuration {
+			delay = maxDuration - spent
+		}
+		if delay > 0 {
+			time.AfterFunc(delay, func() {
```

Comment:
I haven't fully grokked what's going on here, but this seems wrong. Imagine the following scenario: you have a rapid ramp-down, followed by a rapid ramp-up, so that the time "distance" between the two peaks is less than the `gracefulRampDown` value:
```js
{
      type: "variable-looping-vus",
      startVUs: 20,
      stages: [
        { duration: "10s", target: 0 }, // rapid ramp down
        { duration: "10s", target: 20 },  // rapid ramp up
        { duration: "10m", target: 20 },  // stay at 20 VUs for a long time
      ],
      gracefulRampDown: "30s",
    },
```

If an iteration takes 25 seconds, it can start at t=0 and finish at t=25. We shouldn't interrupt it, because `gracefulRampDown` is `30s`. Instead, given that at t=25s, we're back at running 20 VUs, a new iteration should just being in that same VU. Instead, your code seems like it would kill that iteration out of the blue at t=30s.

Also, `time.AfterFunc()` launches its own goroutine, which we probably should avoid.

Reply:
> If an iteration takes 25 seconds, it can start at t=0 and finish at t=25. We shouldn't interrupt it, because `gracefulRampDown` is `30s`.

It wouldn't be interrupted because of the `if rampDowns > 1` check before the `hardStop()` call.

The behavior I see with that script and a `default` function that sleeps for 25s is that 20 VUs are started right away, and while a `gracefulStop()` is called on them immediately, the iterations are allowed to complete and the test stays at 20/20 active VUs essentially disregarding the ramp-down. So 20 iterations are completed at t=25, 40 at t=50, etc. This seems like the correct behavior, no?

> Also, `time.AfterFunc()` launches its own goroutine, which we probably should avoid.

Agreed, it's not ideal, but if I'm understanding [the implementation](https://github.com/golang/go/blob/001fe7f33f1d7aed9e3a047bd8e784bdc103c28c/src/time/sleep.go#L155) and some of the discussion online the goroutine is only created after the timer expires, so it should be more efficient than spawning one directly and sleeping inside it.

I suspect even thousands of VUs shouldn't cause issues, but if this is a blocker I think it can be re-implemented with channels.

Reply:
> The behavior I see with that script and a `default` function that sleeps for 25s is that 20 VUs are started right away, and while a `gracefulStop()` is called on them immediately, the iterations are allowed to complete and the test stays at 20/20 active VUs essentially disregarding the ramp-down. So 20 iterations are completed at t=25, 40 at t=50, etc. This seems like the correct behavior, no?

Yes, that should be what's happening

> It wouldn't be interrupted because of the `if rampDowns > 1` check before the `hardStop()` call.

hmm I still don't fully grok the `rampDowns`, but from what I understand, it's not going to handle some situations nicely. The example I gave seems to work, but this will have a bug:
```
  execution: {
    stages_down_up: {
      type: "variable-looping-vus",
      startVUs: 20,
      stages: [
        { duration: "10s", target: 0 },  // rapid ramp down
        { duration: "10s", target: 3 }, // rapid ramp up to 3 only
        { duration: "10m", target: 3 }, // stay at 3 VUs for a long time
      ],
      gracefulRampDown: "15s", // notice this hard stop
    },
  },
```
notice the `gracefulRampDown` value is different

> > Also, `time.AfterFunc()` launches its own goroutine, which we probably should avoid.
> 
> Agreed, it's not ideal, but if I'm understanding [the implementation](https://github.com/golang/go/blob/001fe7f33f1d7aed9e3a047bd8e784bdc103c28c/src/time/sleep.go#L155) and some of the discussion online the goroutine is only created after the timer expires, so it should be more efficient than spawning one directly and sleeping inside it.
> 
> I suspect even thousands of VUs shouldn't cause issues, but if this is a blocker I think it can be re-implemented with channels.